shap/plots/_style.py

```diff
@@ -0,0 +1,123 @@
+from dataclasses import dataclass, fields, replace
+from typing import TypedDict, Union
+
+import numpy as np
+from typing_extensions import Unpack
+
+from ..utils._exceptions import InvalidOptionError
+from . import colors
+
+# Type hints, adapted from matplotlib.typing
+RGBColorType = Union[tuple[float, float, float], str]
+RGBAColorType = Union[
+    str,  # "none" or "#RRGGBBAA"/"#RGBA" hex strings
+    tuple[float, float, float, float],
+    # 2 tuple (color, alpha) representations, not infinitely recursive
+    # RGBColorType includes the (str, float) tuple, even for RGBA strings
+    tuple[RGBColorType, float],
+    # (4-tuple, float) is odd, but accepted as the outer float overriding A of 4-tuple
+    tuple[tuple[float, float, float, float], float],
+]
+ColorType = Union[RGBColorType, RGBAColorType, np.ndarray]
+
+
+@dataclass(frozen=True)
+class StyleConfig:
+    """A complete set of configuration options for matplotlib-based shap plots."""
+
+    primary_color_positive: ColorType
+    primary_color_negative: ColorType
+    secondary_color_positive: ColorType
+    secondary_color_negative: ColorType
+    hlines_color: ColorType
+    vlines_color: ColorType
+    text_color: ColorType
+    tick_labels_color: ColorType
+
+
+class StyleOptions(TypedDict, total=False):
+    """A TypedDict of partial updates to a style configuration"""
+
+    # Nb. There is some duplication here with the StyleConfig dataclass, but
+    # it's necessary to provide helpful type hints with `typing.Unpack`.
+    # See https://github.com/python/typing/issues/1495
+    primary_color_positive: ColorType
+    primary_color_negative: ColorType
+    secondary_color_positive: ColorType
+    secondary_color_negative: ColorType
+    hlines_color: ColorType
+    vlines_color: ColorType
+    text_color: ColorType
+    tick_labels_color: ColorType
+
+
+_shap_defaults = StyleConfig(
+    primary_color_positive=colors.red_rgb,
+    primary_color_negative=colors.blue_rgb,
+    secondary_color_positive=colors.light_red_rgb,
+    secondary_color_negative=colors.light_blue_rgb,
+    hlines_color="#cccccc",
+    vlines_color="#bbbbbb",
+    text_color="white",
+    tick_labels_color="#999999",
+)
+
+
+def load_default_style() -> StyleConfig:
+    """Load the default style configuration."""
+    # In future, this could allow reading from a persistent config file, like matplotlib rcParams.
+    return _shap_defaults
+
+
+# Singleton instance that determines the current style.
+# CAREFUL! To ensure the correct object is picked up, do not import this directly,
+# but intead access this at runtime with get_style().
+_STYLE = load_default_style()
+
+
+def get_style() -> StyleConfig:
+    """Return all currently active global style configuration options."""
+    return _STYLE
+
+
+def set_style(**options: Unpack[StyleOptions]) -> None:
+    """Set options in the currently active global style configuration."""
+    global _STYLE
+    _STYLE = _apply_options(_STYLE, options)
+
```

Comment:
That's an interesting idea. I do see what you mean about `get_style()` and `set_style()` seeming like they should accept and return the same kind of object. I like your idea, although the type signature is getting a little complex.

How about this as an alternative idea. We could make the entire public API use just the `StyleOptions` objects, by introducing seperate `get_style()` and `get_state()` functions:

```python
# Public API
def get_style() -> StyleOptions: ...
def set_style(**options: Unpack[StyleOptions]) -> None: ...

# For accessing the current runtime values
def get_state() -> StyleConfig: ... 
```

That way, the API is consistent in the sense you can do:
```python
t1 = get_style()
set_style(**t1)
t2 = get_style()
assert t1 == t2
```

Reply:
I pushed a commit with this update, to explore how it looks. What do you think @thatlittleboy ?

Reply:
I would prefer @thatlittleboy's proposal just because output and input parameters do not need an additional transformation. 

I would suggest to use `typing.overload` though, in order to keep this as simple as possible for the user:

```python
@overload
def set_style(style: StyleConfig) -> None:
    ...

@overload
def set_style(**options: Unpack[StyleOptions]) -> None:
    ...

# Implementation of the function
def set_style(style=None, **options):
    if isinstance(style, StyleObject):
        print(f"Setting style using StyleObject: font={style.font}, size={style.size}")
    elif isinstance(options, dict):
        print(f"Setting style using keyword arguments: {options}")
    else:
        raise TypeError("Invalid arguments provided")
```

Reply:
One worry I have about accepting `StyleConfig` as an input is that it's quite susceptible to any breaking changes.

Specifically: if client code creates a `StyleConfig` object, and then we subsequently add a new style option, the client code will be broken as the definition of the dataclass will be incomplete.

So, I think it's safer to have the "set" API designed around the `dict` object which represents a possibility partial set of changes.

Reply:
Wouldn't we just need to set a default for the new option, so that it's backwards compatible? I wouldn't want code to break just to have a new option.

Reply:
OK, I've updated `set_style` to allow passing in a dataclass as a positional parameter.

I still remain a little uncomfortable with this method of setting options, as it seems like a bad conceptual fit. `StyleOptions` represents a "possibly partial set of updates", whereas the dataclass represents a "state of all options" that is guaranteed to be total. The former object seems like the better choice for setting params. 

RE breaking changes: adding default values to all items in the dataclass would indeed help in this regard, but feels a bit messy to me. It is open to misinterpretation: a user might think that `set_style(StyleConfig(foo="bar"))` would simply update the option `foo`, whereas actually this would have the side-effect of setting all other style options as well, potentially overriding anything that was previous set from a user stylesheet. 

So, when we create documentation and examples for this new functionality, I do think we should encourage the setting of options via keyword arguments, and not by instantiating a dataclass directly. Or, put another way, I don't think the DataClass should ever be instantiated directly by a user.

Reply:
PS - this can always be updated in future PRs. I would be keen to try and get a "good enough" first pass implementation merged in, and we are free to add improvements in future.

Reply:
Thanks @connortann  and @CloseChoice ,

> I do think we should encourage the setting of options via keyword arguments, and not by instantiating a dataclass directly. Or, put another way, I don't think the DataClass should ever be instantiated directly by a user.

Agreed.

Let's roll with this implementation first, we can revisit this again in the future once we start playing around more with this (and hopefully also get some user feedback).